File: resources/skype/skypekit-sdk_sdk-4.0.2.138_480749/interfaces/skype/java/client/com/skype/skypekitclient/command/CommandContactList.java
/**
 * Copyright (C) 2010, Skype Limited
 *
 * All intellectual property rights, including but not limited to copyrights,
 * trademarks and patents, as well as know how and trade secrets contained in,
 * relating to, or arising from the internet telephony software of
 * Skype Limited (including its affiliates, "Skype"), including without
 * limitation this source code, Skype API and related material of such
 * software proprietary to Skype and/or its licensors ("IP Rights") are and
 * shall remain the exclusive property of Skype and/or its licensors.
 * The recipient hereby acknowledges and agrees that any unauthorized use of
 * the IP Rights is a violation of intellectual property laws.
 *
 * Skype reserves all rights and may take legal action against infringers of
 * IP Rights.
 *
 * The recipient agrees not to remove, obscure, make illegible or alter any
 * notices or indications of the IP Rights and/or Skype's rights and
 * ownership thereof.
 */

package com.skype.skypekitclient.command;

import com.skype.api.Contact;
import com.skype.api.ContactGroup;

import com.skype.skypekitclient.SkypekitClient;
import com.skype.skypekitclient.utils.Ask;

/**
 *
 */
public class CommandContactList extends ClientCommandInterface
{
    // private static final String TAG = "CommandContactList";

    private SkypekitClient skClient;

    protected CommandContactList(SkypekitClient skypekitClient)
    {
        super();
        skClient = skypekitClient;
    }

    @Override
    public String getName()
    {
        return "ContactList";
    }

    public void executeHelp()
    {
        skClient.out("\n[CONTACTLIST] - l\n"
                + "\tlB - list all buddies\n"
                + "\tlg - list contacts in contact groups\n"
                + "\tlr - rename a contact\n"
                + "\tlp - list contact properties\n"
                + "\tlb - block a contact\n"
                + "\tlu - unblock a contact\n"
                + "\tld - delete a	contact from list\n"
                + "\tla - report abuse\n"
                + "\tlv - store contact avatar\n"
                + "\tlR - refresh profile\n");
    }

    public void execute_p() // list contact properties
    {
        String skypename = Ask.ask("get properties from (skypename) : ");
        Contact contact = find_contact(skypename);
        if (contact == null) {
            skClient.error("Unable to get contact");
            return;
        }
        skClient.out("contact oid " + contact.getOid());

        skClient.out("\t" + "type\t\t\t" + contact.GetIntProperty(Contact.PROPERTY.type));
        skClient.out("\t" + "skypename\t\t" + contact.GetStrProperty(Contact.PROPERTY.skypename));
        skClient.out("\t" + "pstnnumber\t\t" + contact.GetStrProperty(Contact.PROPERTY.pstnnumber));
        skClient.out("\t" + "fullname\t\t" + contact.GetStrProperty(Contact.PROPERTY.fullname));
        skClient.out("\t" + "birthday\t\t" + contact.GetIntProperty(Contact.PROPERTY.birthday));
        skClient.out("\t" + "gender\t\t\t" + contact.GetIntProperty(Contact.PROPERTY.gender));
        skClient.out("\t" + "languages\t\t" + contact.GetStrProperty(Contact.PROPERTY.languages));
        skClient.out("\t" + "country\t\t\t" + contact.GetStrProperty(Contact.PROPERTY.country));
        skClient.out("\t" + "province\t\t" + contact.GetStrProperty(Contact.PROPERTY.province));
        skClient.out("\t" + "city\t\t\t" + contact.GetStrProperty(Contact.PROPERTY.city));
        skClient.out("\t" + "phone_home\t\t" + contact.GetStrProperty(Contact.PROPERTY.phone_home));
        skClient.out("\t" + "phone_office\t\t" + contact.GetStrProperty(Contact.PROPERTY.phone_office));
        skClient.out("\t" + "phone_mobile\t\t" + contact.GetStrProperty(Contact.PROPERTY.phone_mobile));
        skClient.out("\t" + "emails\t\t\t" + contact.GetStrProperty(Contact.PROPERTY.emails));
        skClient.out("\t" + "homepage\t\t" + contact.GetStrProperty(Contact.PROPERTY.homepage));
        skClient.out("\t" + "about\t\t\t" + contact.GetStrProperty(Contact.PROPERTY.about));
        skClient.out("\t" + "avatar_image\t\t" + contact.GetBinProperty(Contact.PROPERTY.avatar_image));
        skClient.out("\t" + "mood_text\t\t" + contact.GetStrProperty(Contact.PROPERTY.mood_text));
        skClient.out("\t" + "rich_mood_text\t\t" + contact.GetStrProperty(Contact.PROPERTY.rich_mood_text));
        skClient.out("\t" + "timezone\t\t" + contact.GetIntProperty(Contact.PROPERTY.timezone));
        skClient.out("\t" + "capabilities\t\t" + contact.GetBinProperty(Contact.PROPERTY.capabilities));
        skClient.out("\t" + "profile_timestamp\t" + contact.GetIntProperty(Contact.PROPERTY.profile_timestamp));
        skClient.out("\t" + "nrof_authed_buddies\t"
                + contact.GetIntProperty(Contact.PROPERTY.nrof_authed_buddies));
        skClient.out("\t" + "ipcountry\t\t" + contact.GetStrProperty(Contact.PROPERTY.ipcountry));
        skClient.out("\t" + "avatar_timestamp\t" + contact.GetIntProperty(Contact.PROPERTY.avatar_timestamp));
        skClient.out("\t" + "mood_timestamp\t\t" + contact.GetIntProperty(Contact.PROPERTY.mood_timestamp));
        skClient.out("\t" + "received_authrequest\t"
                + contact.GetStrProperty(Contact.PROPERTY.received_authrequest));
        skClient.out("\t" + "authreq_timestamp\t" + contact.GetIntProperty(Contact.PROPERTY.authreq_timestamp));
        skClient.out("\t" + "lastonline_timestamp\t"
                + contact.GetIntProperty(Contact.PROPERTY.lastonline_timestamp));
        skClient.out("\t" + "availability\t\t" + contact.GetIntProperty(Contact.PROPERTY.availability));
        skClient.out("\t" + "displayname\t\t" + contact.GetStrProperty(Contact.PROPERTY.displayname));
        skClient.out("\t" + "refreshing\t\t" + contact.GetIntProperty(Contact.PROPERTY.refreshing));
        skClient.out("\t" + "given_authlevel\t\t" + contact.GetIntProperty(Contact.PROPERTY.given_authlevel));
        skClient.out("\t" + "given_displayname\t" + contact.GetStrProperty(Contact.PROPERTY.given_displayname));
        skClient.out("\t" + "assigned_comment\t" + contact.GetStrProperty(Contact.PROPERTY.assigned_comment));
        skClient.out("\t" + "lastused_timestamp\t" + contact.GetIntProperty(Contact.PROPERTY.lastused_timestamp));
        skClient.out("\t" + "authrequest_count\t" + contact.GetIntProperty(Contact.PROPERTY.authrequest_count));
        skClient.out("\t" + "assigned_phone1\t\t" + contact.GetStrProperty(Contact.PROPERTY.assigned_phone1));
        skClient.out("\t" + "assigned_phone1_label\t"
                + contact.GetStrProperty(Contact.PROPERTY.assigned_phone1_label));
        skClient.out("\t" + "assigned_phone2\t\t" + contact.GetStrProperty(Contact.PROPERTY.assigned_phone2));
        skClient.out("\t" + "assigned_phone2_label\t"
                + contact.GetStrProperty(Contact.PROPERTY.assigned_phone2_label));
        skClient.out("\t" + "assigned_phone3\t\t" + contact.GetStrProperty(Contact.PROPERTY.assigned_phone3));
        skClient.out("\t" + "assigned_phone3_label\t"
                + contact.GetStrProperty(Contact.PROPERTY.assigned_phone3_label));
    }

    public void execute_d() // delete a contact from list
    {
        String skypename = Ask.ask("delete contact name: ");
        Contact buddy = find_contact(skypename);
        if (buddy != null) {
            ContactGroup cg = skClient.skype.GetHardwiredContactGroup(ContactGroup.TYPE.ALL_BUDDIES);
            if (cg != null)
                cg.RemoveContact(buddy);
        }
    }

    public Contact find_contact(String skypename)
    {
        if (skClient.notLoggedIn())
            return null;

        ContactGroup cg = skClient.skype.GetHardwiredContactGroup(ContactGroup.TYPE.ALL_KNOWN_CONTACTS);
        Contact[] contacts;

        if ((cg == null) || (null == (contacts = cg.GetContacts()))) {
            skClient.error("Unable to get contact list");
            return null;
        }

        for (int i = 0; i < contacts.length; i++) {
            if ((contacts[i] != null) && (skypename.equals(contacts[i].GetIdentity()))) {
                return contacts[i];
            }
        }

        return null; // No contact with matching skypename
    }

    public void execute_b() // block a contact
    {
        String skypename = Ask.ask("block name: ");
        Contact buddy = find_contact(skypename);
        if (buddy != null) {
            buddy.SetBlocked(true, false);
        }
    }

    public void execute_u() // unblock a contact
    {
        String skypename = Ask.ask("unblock name: ");
        Contact buddy = find_contact(skypename);
        if (buddy != null) {
            buddy.SetBlocked(false, false);
        }
    }

    public void execute_a() // report abuse
    {
        String skypename = Ask.ask("report abuse from: ");
        Contact buddy = find_contact(skypename);
        if (buddy != null) {
            buddy.SetBlocked(true, true);
        }
    }

    public void execute_B() // list all buddies
    {
        if (skClient.notLoggedIn())
            return;

        ContactGroup cg = skClient.skype.GetHardwiredContactGroup(ContactGroup.TYPE.ALL_BUDDIES);
        Contact[] buddies;
        if ((cg == null) || (null == (buddies = cg.GetContacts()))) {
            skClient.error("Unable to get buddies list");
            return;
        }

        skClient.out("buddies = ");

        for (int i = 0; i < buddies.length; i++) {
            Contact buddy = buddies[i];

            skClient.out(buddy.GetStrProperty(Contact.PROPERTY.displayname) + " (" + buddy.GetIdentity() + ") ["
                    + buddy.GetIntProperty(Contact.PROPERTY.availability) + "]");
        }
    }

    public void execute_r() // rename a contact
    {
        String skypename = Ask.ask("rename old name: ");
        Contact buddy = find_contact(skypename);
        if (buddy != null) {
            buddy.GiveDisplayName(Ask.ask("to newname: "));
        }
    }

    public void execute_g() // list contacts in contact groups
    {
        if (skClient.notLoggedIn())
            return;

        ContactGroup.TYPE[] cgTypes = ContactGroup.TYPE.values();
        for (int n = 0; n < cgTypes.length; n++) {
            skClient.out((n + 1) + " - " + cgTypes[n].toString());
        }

        int cnum = Ask.ask_int("select contact group (enter 1-" + cgTypes.length + "): ");

        if ((cnum == 0) || (cnum >= cgTypes.length)) {
            return;
        }

        ContactGroup cg = skClient.skype.GetHardwiredContactGroup(cgTypes[cnum - 1]);

        Contact[] contacts;
        if ((cg == null) || (null == (contacts = cg.GetContacts()))) {
            skClient.error("Unable to get contact list");
            return;
        }

        skClient.out("contacts = ");
        if (contacts.length == 0) {
            skClient.out("none");
            return;
        }

        String identity;
        for (int i = 0; i < contacts.length; i++) {
            Contact contact = contacts[i];
            identity = contact.GetIdentity();
            skClient.out((i + 1) + " - " + identity);
        }
    }

    public void execute_v() // store contact avatar
    {
        String skypename = Ask.ask("skypename: ");
        Contact buddy = find_contact(skypename);
        if (buddy != null) {
            Contact.GetAvatarResult avatar = buddy.GetAvatar();
            if ((avatar != null) && avatar.present) {
                byte[] propavatar = buddy.GetBinProperty(Contact.PROPERTY.avatar_image);
                if (propavatar != null)
                    skClient.out(skypename + " avatar size = " + avatar.avatar.length + " and same as property "
                            + avatar.avatar.equals(propavatar));

            }
            else if (avatar != null) {
                byte[] propavatar = buddy.GetBinProperty(Contact.PROPERTY.avatar_image);
                if (propavatar != null)
                    skClient.out(skypename + " has no avatar using default of size " + avatar.avatar.length
                            + " prop = " + propavatar.length);
            }
            else {
                skClient.error("Failed to get avatar for " + skypename);
            }
        }
    }

    public void execute_R() // refresh profile
    {
        String skypename = Ask.ask("skypename: ");
        Contact contact = find_contact(skypename);
        if (contact != null) {
            contact.RefreshProfile();
        }
    }

}

File: resources/skype/skypekit-sdk_sdk-4.0.2.138_480749/interfaces/skype/cpp_embedded/src/client/command-contactlist.cpp

#include "command-contactlist.h"

void CommandContactList::executeHelp()
{
        printf("\n[CONTACTLIST] - l\n"
               "\tlB - list all buddies\n"
               "\tlg - list contacts in contact groups\n"
               "\tlC - add contact and send auth request\n"
               "\tlr - rename a contact\n"
               "\tlp - list contact properties\n"
               "\tlb - block a contact\n"
               "\tlu - block a contact\n"
               "\tld - delete a	contact from list\n"
               "\tla - report abuse\n"
               "\tlv - store contact avatar\n"
               "\tlR - refresh profile\n"
               "\tlK - show contact capabilities \n"
              );
        fflush(stdout);
}

void CommandContactList::execute_p()
{
        Sid::String skypename = ask("get properties from (skypename) : ");
        CContact::Ref contact = find_contact(skypename);
        if (!contact) {
        	printf("Unable to get contact\n");
                return;        
        }  
        printf("contact oid %d\n", contact->getOID());
         
        List_uint l;
        List_String n;
        l.append(Contact::P_TYPE );
        n.append("type\t\t\t");        
        l.append(Contact::P_SKYPENAME );
        n.append("skypename\t\t");
        l.append(Contact::P_PSTNNUMBER );
        n.append("pstnnumber\t\t");
        l.append(Contact::P_FULLNAME );
        n.append("fullname\t\t");
        l.append(Contact::P_BIRTHDAY );
        n.append("birthday\t\t");
        l.append(Contact::P_GENDER );
        n.append("gender\t\t\t");
        l.append(Contact::P_LANGUAGES );
        n.append("languages\t\t");
        l.append(Contact::P_COUNTRY );
        n.append("country\t\t\t");
        l.append(Contact::P_PROVINCE );
        n.append("province\t\t");
        l.append(Contact::P_CITY );
        n.append("city\t\t\t");
        l.append(Contact::P_PHONE_HOME );
        n.append("phone_home\t\t");
        l.append(Contact::P_PHONE_OFFICE );
        n.append("phone_office\t\t");
        l.append(Contact::P_PHONE_MOBILE );
        n.append("phone_mobile\t\t");
        l.append(Contact::P_EMAILS );
        n.append("emails\t\t\t");
        l.append(Contact::P_HOMEPAGE );
        n.append("homepage\t\t");
        l.append(Contact::P_ABOUT );
        n.append("about\t\t\t");
        l.append(Contact::P_AVATAR_IMAGE );
        n.append("avatar_image\t\t");
        l.append(Contact::P_MOOD_TEXT );
        n.append("mood_text\t\t");
        l.append(Contact::P_RICH_MOOD_TEXT );
        n.append("rich_mood_text\t\t");
        l.append(Contact::P_TIMEZONE );
        n.append("timezone\t\t");
        l.append(Contact::P_CAPABILITIES );
        n.append("capabilities\t\t");
        l.append(Contact::P_PROFILE_TIMESTAMP );
        n.append("profile_timestamp\t");
        l.append(Contact::P_NROF_AUTHED_BUDDIES );
        n.append("nrof_authed_buddies\t");
        l.append(Contact::P_IPCOUNTRY );
        n.append("ipcountry\t\t");
        l.append(Contact::P_AVATAR_TIMESTAMP );
        n.append("avatar_timestamp\t");
        l.append(Contact::P_MOOD_TIMESTAMP );
        n.append("mood_timestamp\t\t");
        l.append(Contact::P_RECEIVED_AUTHREQUEST );
        n.append("received_authrequest\t");
        l.append(Contact::P_AUTHREQ_TIMESTAMP );
        n.append("authreq_timestamp\t");
        l.append(Contact::P_LASTONLINE_TIMESTAMP );
        n.append("lastonline_timestamp\t");
        l.append(Contact::P_AVAILABILITY );
        n.append("availability\t\t");
        l.append(Contact::P_DISPLAYNAME );
        n.append("displayname\t\t");
        l.append(Contact::P_REFRESHING );
        n.append("refreshing\t\t");
        l.append(Contact::P_GIVEN_AUTHLEVEL );
        n.append("given_authlevel\t\t");
        l.append(Contact::P_GIVEN_DISPLAYNAME );
        n.append("given_displayname\t");
        l.append(Contact::P_ASSIGNED_COMMENT );
        n.append("assigned_comment\t");
        l.append(Contact::P_LASTUSED_TIMESTAMP );
        n.append("lastused_timestamp\t");
        l.append(Contact::P_AUTHREQUEST_COUNT );
        n.append("authrequest_count\t");
        l.append(Contact::P_ASSIGNED_PHONE1 );
        n.append("assigned_phone1\t\t");
        l.append(Contact::P_ASSIGNED_PHONE1_LABEL );
        n.append("assigned_phone1_label\t");
        l.append(Contact::P_ASSIGNED_PHONE2 );
        n.append("assigned_phone2\t\t");
        l.append(Contact::P_ASSIGNED_PHONE2_LABEL );
        n.append("assigned_phone2_label\t");
        l.append(Contact::P_ASSIGNED_PHONE3 );
        n.append("assigned_phone3\t\t");
        l.append(Contact::P_ASSIGNED_PHONE3_LABEL );
        n.append("assigned_phone3_label\t");
        
        contact->FetchProps(l);
        Dict_int d = contact->GetProps(l);
        
        for (uint i = 0; i < (uint)l.size(); i++) {
                String value = d.value(l[i]);
                List_String dbg = contact->getPropDebug( l[i], value );
                
                printf("	%s%s\n",(const char*)n[i], (const char*)dbg[2]);
        }
}

void CommandContactList::execute_C()
{
        if (!skype->loggedIn()) {
                printf("You are not logged in\n");
                return;
        }

        CContact::Ref contact;
        if (!skype->GetContact(ask("enter skypename:"), contact)) {
            printf("Cannot get contact\n");
            return;
        }
        if (!contact->SetBuddyStatus(true, true)) {
            printf("Cannot set buddy status\n");
            return;
        }
        if (!contact->SendAuthRequest(ask("enter auth request string:")))
            printf("Cannot send authorization request\n");
}

void CommandContactList::execute_d()
{
        Sid::String skypename = ask("delete contact name: ");
        CContact::Ref buddy = find_contact(skypename);
        if (buddy) {
                ContactGroup::Ref cg;
                bool yes = false;
                skype->GetHardwiredContactGroup(ContactGroup::ALL_BUDDIES,cg);
                if (!cg->CanRemoveContact(yes) || !yes 
                                || !cg->RemoveContact(buddy)) {
                        printf("Can't remove\n");
                        fflush(stdout);
                }
        }
}

Contact::Ref CommandContactList::find_contact(Sid::String& skypename)
{
        if (!skype->loggedIn()) {
                printf("You are not logged in\n");
                return Contact::Ref(0);
        }
        
        ContactGroup::Ref cg;
        Contact::Refs contacts;
        if (!skype->GetHardwiredContactGroup(ContactGroup::ALL_KNOWN_CONTACTS, cg) || !cg->GetContacts(contacts)) {
                printf("Unable to get contact list\n");
                return Contact::Ref(0);
        }
        
        Sid::String identity; 
        for (uint i = 0; i < (uint)contacts.size(); i++) {
                contacts[i]->GetIdentity(identity);
                if (contacts[i] && identity == skypename) {
                        return contacts[i];
                }
        }
        
        return Contact::Ref(0);
}

void CommandContactList::execute_b()
{
        Sid::String skypename = ask("block name: ");
        CContact::Ref buddy = find_contact(skypename);
        if (buddy && !buddy->SetBlocked(true))
                printf("Can't block %s in buddies\n", (const char*)buddy->GetStrProp(Contact::P_DISPLAYNAME));
        fflush(stdout);
}

void CommandContactList::execute_u()
{
        Sid::String skypename = ask("unblock name: ");
        CContact::Ref buddy = find_contact(skypename);
        if (buddy && !buddy->SetBlocked(false))
                printf("Can't unblock %s in buddies\n", (const char*)buddy->GetStrProp(Contact::P_DISPLAYNAME));
        fflush(stdout);
}

void CommandContactList::execute_a()
{
        Sid::String skypename = ask("report abuse from: ");
        CContact::Ref buddy = find_contact(skypename);        
        if (buddy && !buddy->SetBlocked(true,true))
                printf("Can't mark %s in buddies as abusing\n", (const char*)buddy->GetStrProp(Contact::P_DISPLAYNAME));
        fflush(stdout);
}

void CommandContactList::execute_B()
{
        if (!skype->loggedIn()) {
                printf("You are not logged in\n");
                fflush(stdout);
                return;
        }

        ContactGroup::Ref cg;
        Contact::Refs buddies;
        if (!skype->GetHardwiredContactGroup(ContactGroup::ALL_BUDDIES,cg) || !cg->GetContacts(buddies)) {
                printf("Unable to get buddies list\n");
                fflush(stdout);
                return;
        }
        printf("buddies = \n");
        
        List_uint props;
        props.append(Contact::P_DISPLAYNAME);
        props.append(Contact::P_AVAILABILITY);
        skype->FetchMultipleObjectsProps(Contact::MODULE_ID, buddies, props);        
        
        for (uint i = 0; i < (uint)buddies.size(); i++) {
                CContact::Ref buddy = buddies[i];
                Dict_int d = buddy->GetProps(props);
                
                String value = d.value(Contact::P_AVAILABILITY);
                List_String dbg = buddy->getPropDebug(Contact::P_AVAILABILITY, value);
                
                String dname = d.value(Contact::P_DISPLAYNAME);
                printf("%s (%s) [%s]\n",
                       (const char*)dname,
                       (const char*)buddy->skypeName,
                       (const char*)dbg[2]);
        }
        fflush(stdout);
}

void CommandContactList::execute_r()
{
        Sid::String skypename = ask("rename old name: ");
        CContact::Ref buddy = find_contact(skypename);
        if (buddy) {
                buddy->GiveDisplayName(ask("to newname: "));
        }
}

void CommandContactList::execute_g()
{
        if (!skype->loggedIn()) {
                printf("You are not logged in\n");
                fflush(stdout);
                return;
        }
        
        int n = 0;
        while (CONTACTGROUP_TYPE_MAP[n].prop) {
                printf("%d - %s\n", n, CONTACTGROUP_TYPE_MAP[n].prop);
                n++;
        }
        
        char* txt = new char[sizeof("select contact group (enter 1-%d): ")];
        sprintf(txt,"select contact group (enter 1-%d): ", n);
        int cnum = ask_int(txt);
        delete [] txt;
        
        if (cnum >= n) {
                return;
        }
        
        ContactGroup::Ref cg;
        Contact::Refs contacts;
        if (!skype->GetHardwiredContactGroup((ContactGroup::TYPE)CONTACTGROUP_TYPE_MAP[cnum].key, cg) || !cg->GetContacts(contacts)) {
                printf("Unable to get contact list\n");
                return;
        }
        
        printf("contacts = \n");
        if (!contacts.size()) {
                printf("none\n");
                return;
        }
    
        String identity; 
        for (uint i = 0; i < (uint)contacts.size(); i++) {
                CContact::Ref contact = contacts[i];
                contact->GetIdentity(identity);
                printf("%d - %s\n", i+1, (const char*)identity); 
        }       
}


void CommandContactList::execute_v()
{
        Sid::String skypename = ask("skypename: ");
        CContact::Ref buddy = find_contact(skypename);
        if (buddy) {
              Sid::Binary avatar;
              bool present;
              if (buddy->GetAvatar(present, avatar) && present) {
                      Sid::Binary propavatar;
                      buddy->GetPropAvatarImage(propavatar);
                      printf("%s avatar size = %d and same as property %d\n", 
                        (const char*)skypename, (int) avatar.size(), avatar == propavatar); 
                     
              } else {
                     Sid::Binary propavatar;
                     buddy->GetPropAvatarImage(propavatar);
                     printf("%s has no avatar using default of size %d, prop = %d\n", 
                        (const char*)skypename, (int) avatar.size(), (int) propavatar.size()); 
              }
        }
}

void CommandContactList::execute_R() 
{
        Sid::String skypename = ask("skypename: ");
        CContact::Ref contact = find_contact(skypename);
        if (contact) {
                contact->RefreshProfile();        
        }
}

void CommandContactList::execute_K()
{
        Sid::String skypename = ask("get capabilities from (skypename) : ");
        CContact::Ref contact = find_contact(skypename);
        if (!contact) {
                printf("Unable to get contact\n");
                return;        
        } 

        const char* capabilities[] = {
                "VOICEMAIL",
                "SKYPEOUT",
                "SKYPEIN",
                "CAN_BE_SENT_VM",
                "CALL_FORWARD",
                "VIDEO",
                "TEXT",
                "SERVICE_PROVIDER",
                "LARGE_CONFERENCE",
                "COMMERCIAL_CONTACT",
                "PSTN_TRANSFER",
                "TEXT_EVER",
                "VOICE_EVER",
                "MOBILE_DEVICE",
                "PUBLIC_CONTACT"
        };
        const char* capability_status[] = {
                "NOT AVAILABLE",
                "MIXED",
                "EXISTS",
                "FIRST_EXPIRY_WARNING"  /** <30 days*/,
                "SECOND_EXPIRY_WARNING" /** <7 days*/,
                "FINAL_EXPIRY_WARNING"  /** <1 day*/
        };
        
        Contact::CAPABILITYSTATUS s;
        uint expiry_timestamp = 0;
        for (int c = 0; c <= Contact::CAPABILITY_PUBLIC_CONTACT; c++) {
                if (!contact->GetCapabilityStatus((Contact::CAPABILITY) c, s, expiry_timestamp == 1)) {
                        printf("error when getting capability\n");
                        fflush(stdout);
                        return;
                }
                if (s > Contact::CAPABILITY_EXISTS)
                        printf("capability %s status is %s (expires on %d)\n", capabilities[c], capability_status[s], expiry_timestamp);
                else
                        printf("capability %s status is %s\n", capabilities[c], capability_status[s]);
                
                fflush(stdout);
        }
}







